daisy_workflows/image_import/ubuntu/translate_ubuntu.sh

```diff
@@ -15,108 +15,36 @@
 
 set -x
 
+# Download translate_el.py
 URL="http://metadata/computeMetadata/v1/instance/attributes"
-UBU_RELEASE="$(curl -f -H Metadata-Flavor:Google ${URL}/ubuntu_release)"
-INSTALL_GCE="$(curl -f -H Metadata-Flavor:Google ${URL}/install_gce_packages)"
-MNT="/mnt/imported_disk"
-
-# Mount the imported disk.
-mkdir ${MNT}
-if [ -b /dev/sdb1 ]; then
-  echo "Trying to mount /dev/sdb1"
-  mount /dev/sdb1 ${MNT}
-  if [ $? -ne 0 ]; then
-    echo "TranslateFailed: Unable to mount imported disk."
-    exit 1
-  fi
-else
-  echo "Trying to mount /dev/sdb"
-  mount /dev/sdb ${MNT}
-  if [ $? -ne 0 ]; then
-    echo "TranslateFailed: Unable to mount imported disk."
-    exit 1
-  fi
-fi
-
-# Setup DNS for chroot.
-for f in proc sys dev run; do
-  mount -o bind /$f ${MNT}/$f
-done
-mkdir -p /run/resolvconf
-cp /etc/resolv.conf /run/resolvconf/resolv.conf
-
-# Install GCE packages if requested.
-if [[ "${INSTALL_GCE}" == "true" ]]; then
-  chroot ${MNT} apt-get update
-  echo "Installing cloud-init."
-  chroot ${MNT} apt-get install -y --no-install-recommends cloud-init
-
-  # Try to remove azure or aws configs so cloud-init has a chance.
-  rm -f ${MNT}/etc/cloud/cloud.cfg.d/*azure*
-  rm -f ${MNT}/etc/cloud/cloud.cfg.d/*waagent*
-  rm -f ${MNT}/etc/cloud/cloud.cfg.d/*walinuxagent*
-  rm -f ${MNT}/etc/cloud/cloud.cfg.d/*aws*
-  rm -f ${MNT}/etc/cloud/cloud.cfg.d/*amazon*
-
-  # Remove Azure agent.
-  chroot ${MNT} apt-get remove -y -f waagent walinuxagent
-
-  cat > ${MNT}/etc/apt/sources.list.d/partner.list <<EOF
-# Enabled for Google Cloud SDK
-deb http://archive.canonical.com/ubuntu ${UBU_RELEASE} partner
-EOF
-
-  cat > ${MNT}/etc/cloud/cloud.cfg.d/91-gce-system.cfg <<EOF
-datasource_list: [ GCE ]
-system_info:
-   package_mirrors:
-     - arches: [i386, amd64]
-       failsafe:
-         primary: http://archive.ubuntu.com/ubuntu
-         security: http://security.ubuntu.com/ubuntu
-       search:
-         primary:
-           - http://%(region)s.gce.archive.ubuntu.com/ubuntu/
-           - http://%(availability_zone)s.gce.clouds.archive.ubuntu.com/ubuntu/
-           - http://gce.clouds.archive.ubuntu.com/ubuntu/
-         security: []
-     - arches: [armhf, armel, default]
-       failsafe:
-         primary: http://ports.ubuntu.com/ubuntu-ports
-         security: http://ports.ubuntu.com/ubuntu-ports
+GCS_SOURCE_PATH="$(curl -f -H Metadata-Flavor:Google ${URL}/daisy-sources-path)"
+gsutil cp "${GCS_SOURCE_PATH}/translate_ubuntu.py" /tmp/translate_ubuntu.py
+
+# Install dependencies
+DEBIAN_FRONTEND=noninteractive apt-get -y install python-guestfs python-pycurl tinyproxy
+
+# cloud-init command uses GCE API, as libguestfs executes inside qemu, it uses a
```

Comment:
How does this work? GCE doesn't support nested qemu.

Reply:
Qemu can be used to run an appliance instead of a full VM, there is a nice explanation here https://rwmj.wordpress.com/2011/07/07/how-does-libguestfs-live-work/
And the official docs http://libguestfs.org/guestfs-internals.1.html

Reply:
Ahh ok. Interesting.